tracing.go

```diff
@@ -84,53 +87,133 @@ func (s Span) ParentID() uint64 { return s.parentID }
 // SpanID returns the id of this specific RPC
 func (s Span) SpanID() uint64 { return s.spanID }
 
-// EnableTracing controls whether tracing is enabled for this context
-func (s *Span) EnableTracing(enabled bool) {
-	if enabled {
-		s.flags |= tracingFlagEnabled
-	} else {
-		s.flags &= ^tracingFlagEnabled
+// Flags returns flags bitmap. Interpretation of the bits is up to the tracing system.
+func (s Span) Flags() byte { return s.flags }
+
+// SetTraceID sets traceID
+func (s *Span) SetTraceID(traceID uint64) { s.traceID = traceID }
+
+// SetSpanID sets spanID
+func (s *Span) SetSpanID(spanID uint64) { s.spanID = spanID }
+
+// SetParentID sets parentID
+func (s *Span) SetParentID(parentID uint64) { s.parentID = parentID }
+
+// SetFlags sets flags
+func (s *Span) SetFlags(flags byte) { s.flags = flags }
+
+// CurrentSpan extracts OpenTracing Span from the Context, and if found tries to
+// extract zipkin-style trace/span IDs from it using ZipkinSpanFormat carrier.
+func CurrentSpan(ctx context.Context) *Span {
+	if sp := opentracing.SpanFromContext(ctx); sp != nil {
+		span := &Span{}
+		span.initFromOpenTracing(sp)
+		return span
 	}
+	return nil
 }
 
-// TracingEnabled checks whether tracing is enabled for this context
-func (s Span) TracingEnabled() bool { return (s.flags & tracingFlagEnabled) == tracingFlagEnabled }
+// initFromOpenTracing initializes Span fields from an OpenTracing Span,
+// assuming the tracing implementation supports Zipkin-style span IDs.
+func (s *Span) initFromOpenTracing(span opentracing.Span) {
+	if err := span.Tracer().Inject(span.Context(), ZipkinSpanFormat, s); err != nil {
+		s.initRandom()
+	}
+}
 
-// NewChildSpan begins a new child span in the provided Context
-func (s Span) NewChildSpan() *Span {
-	childSpan := &Span{
-		traceID:  s.traceID,
-		parentID: s.spanID,
-		flags:    s.flags,
+func (c *Connection) startOutboundSpan(ctx context.Context, serviceName, methodName string, call *OutboundCall, startTime time.Time) opentracing.Span {
+	var parent opentracing.SpanContext // ok to be nil
+	if s := opentracing.SpanFromContext(ctx); s != nil {
+		parent = s.Context()
 	}
-	if s.spanID == 0 {
-		childSpan.spanID = childSpan.traceID
-	} else {
-		childSpan.spanID = uint64(traceRng.Int63())
+	span := c.Tracer().StartSpan(
+		serviceName+"::"+methodName,
```

Comment:
the `serviceName` here is not the Thrift service name, it's the routing service name, so for example in moe, the values would be:
`serviceName = "moe"`
`methodName = "Stooges::nyuck"`

If you append them, the spans will become `moe::Stooges::nyuck` which I don't think is what we want.